buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestoneNextVersionDueTodayTask.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2019-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.gradle.github.milestones;
+
+import org.gradle.api.Action;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.TaskAction;
+
+import org.springframework.gradle.github.RepositoryRef;
+
+public class GitHubMilestoneNextVersionDueTodayTask extends DefaultTask {
+
+	@Input
+	private RepositoryRef repository = new RepositoryRef();
+
+	@Input
+	@Optional
+	private String nextVersion;
+
+	@Input
+	@Optional
+	private String gitHubAccessToken;
+
+	private GitHubMilestoneApi milestones = new GitHubMilestoneApi();
+
+	@TaskAction
+	public void checkReleaseDueToday() {
+		if (this.nextVersion == null) {
+			String currentVersion = (String) getProject().getVersion();
+			this.nextVersion = this.milestones.calculateNextReleaseMilestone(repository, currentVersion);
+			System.out.println("Next version was not provided. Calculated next version as " + this.nextVersion);
+		}
+		boolean milestoneDueToday = this.milestones.isMilestoneDueToday(this.repository, this.nextVersion);
+		if (!milestoneDueToday) {
+			throw new IllegalStateException("The milestone with the title " + this.nextVersion
+					+ " in the repository " + this.repository + " is not due yet");
+		}
+		System.out.println("The milestone with the title " + this.nextVersion + " in the repository "
+				+ this.repository + " is due today");
```

Comment:
After discussion, I did a quick test, and it looks like if we only output the next version if `milestoneDueToday` is true, and (possibly) a blank line otherwise, we can capture that output with the `-q` flag.

For example:

```sh
nextVersion=$(./gradlew -q gitHubCheckMilestoneDueToday)
if [[ "$nextVersion" != "" ]]
then
    echo "Starting release for next version"
fi
```

**Note:** We could also set a task output and check that in a task further down, if we want to avoid the bash if-statement.

Perhaps we could split this into two tasks:

1. `gitHubCheckMilestoneDueToday` fails the build if the given/calculated version is not due today (can be used as a guard against accidentally releasing at the wrong time)
2. `gitHubCalculateNextReleaseMilestoneForToday` (or similar) which simply outputs the next version, or nothing/blank line

Reply:
I agree it makes sense to split it up into 2 tasks. I was just talking to Rob and he suggested looking into output / input files, since that is a standard way to share information between tasks. I'll start with these updates over the next few days.

Reply:
As you suggested I split it up into two tasks:
1. `gitHubNextReleaseMilestone` which outputs the next version to a file called `next-release.yml`
2. `gitHubCheckMilestoneDueToday` which depends on `gitHubNextReleaseMilestone` and uses the generated file `next-release.yml` as input and fails the build if the given/calculated version is not due today